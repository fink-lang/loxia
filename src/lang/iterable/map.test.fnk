{describe, it, expect, to_match_snapshot} = import '@fink/jest/test.fnk'

{fink2js} = import '../../testing/generate.fnk'


describe 'map', fn:
  it 'compiles single line', fn:
    expect
      fink2js '
        map item: item * 2
      '
      to_match_snapshot


  it 'compiles multi line', fn:
    expect
      fink2js '
        map item:
          ni = foo item
          ni + 2
      '
      to_match_snapshot


  it 'compiles single line with default value', fn:
    expect
      fink2js '
        map item=123: item * 2
      '
      to_match_snapshot


  it 'compiles single line with destructured obj', fn:
    expect
      fink2js '
        map {item}: item * 2
      '
      to_match_snapshot


  it 'compiles single line with destructured list', fn:
    expect
      fink2js '
        map [x, y]: x + y
      '
      to_match_snapshot


  it 'compiles as flat map', fn:
    expect
      fink2js '
        map [x, y]:
          ...[x, y]
      '
      to_match_snapshot


  it 'compiles with foo', fn:
    expect
      fink2js '
        map {foo}:
          match foo:
            bar: spam
      '
      to_match_snapshot


  it 'compiles with acc', fn:
    expect
      fink2js '
        map {foo}, acc=0:
          [foo + acc, acc+1]
      '
      to_match_snapshot

    expect
      fink2js '
        map {foo}, acc=0:
          ...[[acc, foo], acc+1]
      '
      to_match_snapshot


  it 'compiles destructuring', fn:
    expect
      fink2js '
        map [x, ...rest]:
         rest

        map [...ignored, last]:
         last
      '
      to_match_snapshot



describe 'map async', fn:
  it 'compiles', fn:
    expect
      fink2js '
        map item:
          await item * 2
      '
      to_match_snapshot