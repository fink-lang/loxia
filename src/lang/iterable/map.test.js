import {fink2js} from '../../testing';


describe('map', ()=> {
  it('compiles single line', ()=> {
    expect(
      fink2js(`
        map item: item * 2
      `)
    ).toMatchSnapshot();
  });


  it('compiles multi line', ()=> {
    expect(
      fink2js(`
        map item:
          item * 2
      `)
    ).toMatchSnapshot();
  });


  it('compiles single line with default value', ()=> {
    expect(
      fink2js(`
        map item=123: item * 2
      `)
    ).toMatchSnapshot();
  });


  it('compiles single line with destructured obj', ()=> {
    expect(
      fink2js(`
        map {item}: item * 2
      `)
    ).toMatchSnapshot();
  });


  it('compiles single line with destructured array', ()=> {
    expect(
      fink2js(`
        map [x, y]: x + y
      `)
    ).toMatchSnapshot();
  });


  it('compiles as flat map', ()=> {
    expect(
      fink2js(`
        map [x, y]:
          ...[x, y]
      `)
    ).toMatchSnapshot();
  });
});


describe('flat_map', ()=> {

  it('compiles', ()=> {
    expect(
      fink2js(`
        flat_map item: [spam + item]

        flat_map item: [spam]
      `)
    ).toMatchSnapshot();
  });

});


