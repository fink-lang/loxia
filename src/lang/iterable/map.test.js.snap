// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flat_map compiles 1`] = `
"(function* map(ˆitems_1) {
  for (const item of ˆitems_1) {
    const ˆresult_2 = [spam + item];
    yield* ˆresult_2;
  }
});

(function* map(ˆitems_3) {
  for (const item of ˆitems_3) {
    const ˆresult_4 = [spam];
    yield* ˆresult_4;
  }
});

Object.assign(module.exports, {});"
`;

exports[`map compiles as flat map 1`] = `
"(function* map(ˆitems_1) {
  for (const [x, y] of ˆitems_1) {
    const ˆresult_2 = [x, y];
    yield* ˆresult_2;
  }
});

Object.assign(module.exports, {});"
`;

exports[`map compiles multi line 1`] = `
"(function* map(ˆitems_1) {
  for (const item of ˆitems_1) {
    const ˆresult_2 = item * 2;
    yield ˆresult_2;
  }
});

Object.assign(module.exports, {});"
`;

exports[`map compiles single line 1`] = `
"(function* map(ˆitems_1) {
  for (const item of ˆitems_1) {
    const ˆresult_2 = item * 2;
    yield ˆresult_2;
  }
});

Object.assign(module.exports, {});"
`;

exports[`map compiles single line with default value 1`] = `
"(function* map(ˆitems_1) {
  for (const item = 123 of ˆitems_1) {
    const ˆresult_2 = item * 2;
    yield ˆresult_2;
  }
});

Object.assign(module.exports, {});"
`;

exports[`map compiles single line with destructured array 1`] = `
"(function* map(ˆitems_1) {
  for (const [x, y] of ˆitems_1) {
    const ˆresult_2 = x + y;
    yield ˆresult_2;
  }
});

Object.assign(module.exports, {});"
`;

exports[`map compiles single line with destructured obj 1`] = `
"(function* map(ˆitems_1) {
  for (const {
    \\"item\\": item
  } of ˆitems_1) {
    const ˆresult_2 = item * 2;
    yield ˆresult_2;
  }
});

Object.assign(module.exports, {});"
`;
